### PR TITLE
Enable property functions for line-width

### PR DESCRIPTION
**Pull request by lbud**
Ref #3045 

The challenge in implementing data-driven `line-width` has been that `line-dasharray` is scaled by `line-width`, so dasharray calculations require the value of a line's width at floor(zoom). This PR accomplishes data-driven `line-width` by moving that calculation to the shader and adding an additional, sort of phantom paint attribute that calculates the value of line-width at floor(zoom). This required a bit of special casing in program_configuration but given the existing situation seemed to me the best option. This would nullify #4158.  @ansis @jfirebaugh @nickidlugash @anandthakker — does this seem reasonable?
 
I verified manually with a matrix of 
`{ line-width: [constant, zoom, property, zoom-and-property] }` 
x 
`{ line-dasharray: [none, constant, zoom] }` 
and added render tests here to the same effect (the naming in the render suite is a bit verbose 💁).

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] manually test the debug page
 - [x] gl-native companion PR -> https://github.com/mapbox/mapbox-gl-native/pull/9214/

### CODE REVIEW

File: src/data/bucket/line_bucket.js [hunk 1]
```diff
@@ -50,6 +50,8 @@ const lineInterface = {
         {property: 'line-opacity', multiplier: 10, type: 'Uint8'},
         {property: 'line-gap-width', multiplier: 10, type: 'Uint8', name: 'a_gapwidth'},
         {property: 'line-offset', multiplier: 1, type: 'Int8'},
+        {property: 'line-width', multiplier: 10, type: 'Uint8', name: 'a_width'},
+        {property: 'line-width', multiplier: 10, type: 'Uint8', name: 'a_floorwidth'},
```
Comment (ansis) [inline src/data/bucket/line_bucket.js]: This approach looks great!

I wonder if adding a `floorZoom: true` to the property definition here and using that to making the `if (name === 'a_floorwidth')` exceptions less exceptional would be good. Or if that would just hide the exception in a bad way.

Comment (lbud) [inline src/data/bucket/line_bucket.js]: 👍 I like that idea

**Comment by mollymerp** [comment]
@lbud awesome 🕵️‍♀️ !!

**Comment by keyofj** [comment]
Thanks for making this a priority. Dynamic line-width is an important feature.

File: test/integration/render-tests/line-dasharray/function--line-width-composite-function/style.json [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+    52.499167
+  ],
+  "zoom": 14,
+  "sources": {
+    "mapbox": {
+      "type": "vector",
+      "maxzoom": 14,
+      "tiles": [
+        "local://tiles/{z}-{x}-{y}.mvt"
+      ]
```
Comment (anandthakker) [inline test/integration/render-tests/line-dasharray/function--line-width-composite-function/style.json]: I'd be in favor of simplifying these render tests using some bare-bones GeoJSON, as that'll make em easier to understand/assess.

**Comment by lbud** [comment]
@anandthakker thanks for the suggestion — simplified geojson fixtures in  https://github.com/mapbox/mapbox-gl-js/pull/4773/commits/467ed46c71aea0dd638140b8618e4e955623ea93

**Pull request merged**